conans/model/workspace.py

```diff
@@ -1,155 +1,121 @@
-        if os.path.exists(path):
-            return Workspace(path, install_folder)
-        parent = os.path.dirname(folder)
-        if parent and parent != folder:
-            return Workspace.get_workspace(parent, install_folder)
-        return None
-
-    def generate(self):
-        if self._generator == "cmake":
-            template = """# conanws
-cmake_minimum_required(VERSION 3.3)
-project({name} CXX)
-
-"""
-            cmake = template.format(name=self._name)
-            for _, workspace_package in self._workspace_packages.items():
-                build_folder = workspace_package.build_folder
-                build_folder = build_folder.replace("\\", "/")
-                cmake += 'add_subdirectory(%s "%s")\n' % (workspace_package.local_cmakedir, build_folder)
-            cmake_path = os.path.join(self._base_folder, "CMakeLists.txt")
-            if os.path.exists(cmake_path) and not load(cmake_path).startswith("# conanws"):
-                raise ConanException("Can't generate CMakeLists.txt, will overwrite existing one")
+
+    def generate(self, cwd, graph):
+        editables = {node.ref: node.conanfile for node in graph.nodes
+                     if node.recipe == RECIPE_EDITABLE}
+        if self._ws_generator == "cmake":
+            cmake = ""
+            add_subdirs = ""
+            for node in graph.ordered_iterate():
+                if node.recipe != RECIPE_EDITABLE:
+                    continue
+                ref = node.ref
+                ws_pkg = self._workspace_packages[ref]
+                layout = self._cache.package_layout(ref)
+                editable = layout.editable_cpp_info()
+                conanfile = editables[ref]
+                build = editable.folder(ref, "build_folder", conanfile.settings, conanfile.options)
+                src = editable.folder(ref, "source_folder", conanfile.settings, conanfile.options)
+                if src:
+                    src = os.path.join(ws_pkg.root_folder, src).replace("\\", "/")
+                    cmake += 'set(PACKAGE_%s_SRC "%s")\n' % (ref.name, src)
+                if build:
+                    build = os.path.join(ws_pkg.root_folder, build).replace("\\", "/")
+                    cmake += 'set(PACKAGE_%s_BUILD "%s")\n' % (ref.name, build)
+
+                if src and build:
+                    add_subdirs += ('    add_subdirectory(${PACKAGE_%s_SRC} ${PACKAGE_%s_BUILD})\n'
+                                    % (ref.name, ref.name))
+            if add_subdirs:
+                cmake += "macro(conan_workspace_subdirectories)\n"
+                cmake += add_subdirs
+                cmake += "endmacro()"
+            cmake_path = os.path.join(cwd, "conanworkspace.cmake")
             save(cmake_path, cmake)
 
-    def __init__(self, path, install_folder):
-        self._install_folder = install_folder
-        self._generator = None
-        self._name = "ConanWorkspace"
+    def __init__(self, path, cache):
+        self._cache = cache
```

Comment:
I think we can avoid passing the `cache` object here. It is used just in the `Workspace::generate` function (ok, also to get the `cache.conan_folder`, but it can be easily substituted) from the `conan_api` where we already know which packages are in _editable mode_, so we can get there the `package_layouts` we are going to use and pass them to this class.

Reply:
leaving this for later